runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/graph/OutputDeduplicator.java

```diff
@@ -0,0 +1,345 @@
+      PCollectionNode alreadyDeduplicated =
+          unzippedOutputs.put(duplicateOutput.getId(), PipelineNode.pCollection(id, partial));
+      checkArgument(alreadyDeduplicated == null, "a duplicate should only appear once per stage");
+    }
+    return unzippedOutputs;
+  }
+
+  /**
+   * Returns an {@link ExecutableStage} where all of the {@link PCollectionNode PCollections}
+   * matching the original are replaced with the introduced partial {@link PCollection} in all
+   * references made within the {@link ExecutableStage}.
+   */
+  private static ExecutableStage deduplicateStageOutput(
+      ExecutableStage stage, Map<String, PCollectionNode> originalToPartial) {
+    Collection<PTransformNode> updatedTransforms = new ArrayList<>();
+    for (PTransformNode transform : stage.getTransforms()) {
+      PTransform updatedTransform = updateOutputs(transform.getTransform(), originalToPartial);
+      updatedTransforms.add(PipelineNode.pTransform(transform.getId(), updatedTransform));
+    }
+    Collection<PCollectionNode> updatedOutputs = new ArrayList<>();
+    for (PCollectionNode output : stage.getOutputPCollections()) {
+      updatedOutputs.add(originalToPartial.getOrDefault(output.getId(), output));
+    }
+    RunnerApi.Components updatedStageComponents =
+        stage
+            .getComponents()
+            .toBuilder()
+            .clearTransforms()
+            .putAllTransforms(
+                updatedTransforms
+                    .stream()
+                    .collect(Collectors.toMap(PTransformNode::getId, PTransformNode::getTransform)))
+            .putAllPcollections(
+                originalToPartial
+                    .values()
+                    .stream()
+                    .collect(
+                        Collectors.toMap(PCollectionNode::getId, PCollectionNode::getPCollection)))
+            .build();
+    return ImmutableExecutableStage.of(
+        updatedStageComponents,
+        stage.getEnvironment(),
+        stage.getInputPCollection(),
+        stage.getSideInputPCollections(),
+        updatedTransforms,
+        updatedOutputs);
+  }
+
+  /**
+   * Returns a {@link PTransform} like the input {@link PTransform}, but with each output to {@code
+   * originalPCollection} replaced with an output (with the same local name) to {@code
+   * newPCollection}.
+   */
+  private static PTransform updateOutputs(
+      PTransform transform, Map<String, PCollectionNode> originalToPartial) {
+    PTransform.Builder updatedTransformBuilder = transform.toBuilder();
+    for (Map.Entry<String, String> output : transform.getOutputsMap().entrySet()) {
+      if (originalToPartial.containsKey(output.getValue())) {
+        updatedTransformBuilder.putOutputs(
+            output.getKey(), originalToPartial.get(output.getValue()).getId());
+      }
+    }
+    return updatedTransformBuilder.build();
+  }
+
+  @AutoValue
+  abstract static class StageOrTransform {
```

Comment:
It seems a lot of this code (and possibly code elsewhere) could be simplified if a ExecutableStage was an instance of (or could be wrapped in) PTransformNode. Is there any reason that isn't the case?

Reply:
OK, I'll defer to you on making this tradeoff.